subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/GetInputFilesVisitor.java

```diff
@@ -39,11 +40,11 @@ public GetInputFilesVisitor(String ownerDisplayName, FileCollectionFactory fileC
     }
 
     @Override
-    public void visitInputFileProperty(final String propertyName, boolean optional, boolean skipWhenEmpty, Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
+    public void visitInputFileProperty(final String propertyName, boolean optional, boolean skipWhenEmpty, @Nullable Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
         FileCollection actualValue = FileParameterUtils.resolveInputFileValue(fileCollectionFactory, filePropertyType, value);
         specs.add(new DefaultInputFilePropertySpec(
             propertyName,
-            fileNormalizer,
+            FileParameterUtils.normalizerOrDefault(fileNormalizer),
```

Comment:
Instead of passing in `@Nullable` all the way to here, can we do this transformation somewhere earlier?

Reply:
Doing the transformation earlier would mean in the annotation handler. We currently don't have a way to report problems in the annotation handler itself. The annotation handler also does not have the means to decide whether a transform is cacheable or not.

Do you think it makes sense to do the validation in the annotation handler instead? For tasks, that would also mean we need to extract the validation from the annotation handler, since we do task validation later and also for inputs registered at runtime. We can handle those problems as part of https://github.com/gradle/build-cache/issues/1298.

Reply:
Let's do that.